File: source/server/src/static.ts
/**
 * 托管前端产物（web/dist）。
 *
 * - 有产物：静态文件 + SPA 回退（任何非 /v1、非 /healthz 的 GET 都返回 index.html）；
 * - 没有产物：直接不注册，只提供 API——本地用 `npm run dev -w @daybook/web`（Vite 代理 /v1）时就是这样。
 *
 * 因此镜像自包含：一个容器同时提供页面与 API，反代只需指向 8090（见计划 §7.1）。
 */
import fastifyStatic from '@fastify/static';
import { existsSync } from 'node:fs';
import { join } from 'node:path';

import type { FastifyInstance } from 'fastify';

const ONE_HOUR_MS = 60 * 60 * 1000;

/** 看起来像静态资源（带常见扩展名）的路径 */
const ASSET_PATTERN =
  /\.(?:js|mjs|cjs|css|map|json|webmanifest|png|jpe?g|gif|svg|ico|webp|avif|woff2?|ttf|otf|txt|xml|wasm)$/i;

export async function registerWebApp(app: FastifyInstance, webDistDir: string): Promise<boolean> {
  if (!existsSync(join(webDistDir, 'index.html'))) return false;

  await app.register(fastifyStatic, {
    root: webDistDir,
    index: ['index.html'],
    cacheControl: true,
    maxAge: ONE_HOUR_MS,
    setHeaders(response, path) {
      // 注意：@fastify/static v8 传进来的是 Fastify Reply（用 .header），
      // 早期版本传原生 ServerResponse（用 .setHeader）——两种都兼容，写错会直接崩进程。
      const target = response as unknown as {
        header?: (name: string, value: string) => unknown;
        setHeader?: (name: string, value: string) => unknown;
      };
      const setHeader = (name: string, value: string): void => {
        if (typeof target.header === 'function') {
          target.header(name, value);
          return;
        }
        if (typeof target.setHeader === 'function') {
          target.setHeader(name, value);
        }
      };

      // 文件名带内容哈希的构建产物可以长期缓存
      if (path.includes(`${join('assets')}`)) {
        setHeader('cache-control', 'public, max-age=31536000, immutable');
      }
      // Service Worker 不能被长缓存，否则更新发不出去
      if (path.endsWith('sw.js')) {
        setHeader('cache-control', 'no-cache');
      }
    },
  });

  app.setNotFoundHandler((request, reply) => {
    const url = (request.url.split('?')[0] ?? '/').replace(/\\/g, '/');
    const isApi = url.startsWith('/v1/') || url === '/healthz';
    // 静态资源找不到就该 404。早期版本一律回 index.html：发版后旧页面引用到已被清掉的
    // /assets/xxx.js 时会拿到 200 的 HTML，浏览器报 MIME 错误 → 白屏，且极难排查。
    const looksLikeAsset = url.startsWith('/assets/') || ASSET_PATTERN.test(url);

    if (!isApi && !looksLikeAsset && request.method === 'GET') return reply.sendFile('index.html');
    return reply.code(404).send({ error: 'not_found' });
  });

  return true;
}

File: source/server/api/static.go
package api

import (
	"mime"
	"net/http"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/pocketbase/pocketbase/core"
)

// assetPattern 判断"看起来像静态资源"的路径（带常见扩展名）。
// 与 Node 版 static.ts 的正则一致。
var assetPattern = regexp.MustCompile(
	`(?i)\.(?:js|mjs|cjs|css|map|json|webmanifest|png|jpe?g|gif|svg|ico|webp|avif|woff2?|ttf|otf|txt|xml|wasm)$`,
)

// RegisterStatic 托管前端产物并做 SPA 回退。
//
// 返回 false 表示没有产物——本地开发时就是这个状态（前端跑 Vite dev server，
// 由它代理 /v1），此时只提供 API。
func RegisterStatic(event *core.ServeEvent, publicDir string) bool {
	if _, err := os.Stat(filepath.Join(publicDir, "index.html")); err != nil {
		return false
	}

	// 通配路由在 API 路由之后注册：Echo 先匹配已注册的精确路由，
	// 所以 /v1/… 与 /_/ 不会被这里吃掉。
	event.Router.GET("/{path...}", func(request *core.RequestEvent) error {
		// path.Clean 会消掉 ../，配合 os.DirFS/os.ReadFile 的路径拼接，
		// 请求方无法用 "..%2f" 之类的手段跑出产物目录。
		cleaned := path.Clean("/" + strings.TrimPrefix(request.Request.PathValue("path"), "/"))
		name := strings.TrimPrefix(cleaned, "/")

		if name == "" {
			return serveIndex(request, publicDir)
		}
		if data, err := os.ReadFile(filepath.Join(publicDir, filepath.FromSlash(name))); err == nil {
			return writeFile(request, name, data)
		}

		// 找不到文件。这里必须把三类路径分开处理——其中两条是 Node 版
		// 踩过坑之后特意加上的防护，理由写在各自的分支里。
		if isAPIPath(cleaned) || assetPattern.MatchString(cleaned) {
			return fail(request, http.StatusNotFound, "not_found")
		}
		if request.Request.Method != http.MethodGet {
			return fail(request, http.StatusNotFound, "not_found")
		}
		// 其余 GET 当作前端路由（/calendar、/day/2026-09-13 …）交给 SPA。
		return serveIndex(request, publicDir)
	})

	return true
}

// isAPIPath 判断是不是 API 路径。
//
// 为什么要单独判：API 找不到时返回 index.html 会让前端把 HTML 当 JSON 解析，
// 报出一个和真实原因毫无关系的解析错误。
func isAPIPath(cleaned string) bool {
	if cleaned == "/healthz" {
		return true
	}
	return cleaned == "/v1" || strings.HasPrefix(cleaned, "/v1/")
}

// applyCachePolicy 沿用 Node 版的缓存策略。
//
// 关键是 sw.js 与 index.html：Service Worker 一旦被长缓存，新版本就再也发不出去
// （用户浏览器一直拿着旧的 sw.js 和旧的资源清单）；index.html 引用的资源文件名
// 带哈希，它自己被缓存住的话用户会一直加载上一次的产物清单。
//
// 这里比 Node 版多一条：Node 版对直接请求 /index.html 也是 1 小时缓存，
// 只有 SPA 回退那条路径才设 no-cache。同一个文件两种缓存策略是隐患，
// 所以这里统一按 no-cache 处理。
func applyCachePolicy(header http.Header, name string) {
	base := path.Base(name)
	switch {
	case strings.HasPrefix(name, "assets/"):
		// vite 产物文件名带内容哈希，内容变了文件名就变 → 可以永久缓存
		header.Set("Cache-Control", "public, max-age=31536000, immutable")
	case base == "sw.js" || base == "index.html":
		header.Set("Cache-Control", "no-cache")
	default:
		header.Set("Cache-Control", "public, max-age=3600")
	}
}

func writeFile(request *core.RequestEvent, name string, data []byte) error {
	header := request.Response.Header()
	contentType := mime.TypeByExtension(path.Ext(name))
	if contentType == "" {
		contentType = http.DetectContentType(data)
	}
	header.Set("Content-Type", contentType)
	applyCachePolicy(header, name)
	request.Response.WriteHeader(http.StatusOK)
	_, err := request.Response.Write(data)
	return err
}

// serveIndex 返回 index.html。
//
// 刻意用 no-cache：它引用的资源文件名带哈希，发版后如果 index.html 被缓存住，
// 用户会一直加载上一次的产物清单。
func serveIndex(request *core.RequestEvent, publicDir string) error {
	data, err := os.ReadFile(filepath.Join(publicDir, "index.html"))
	if err != nil {
		return fail(request, http.StatusNotFound, "not_found")
	}
	header := request.Response.Header()
	header.Set("Content-Type", "text/html; charset=utf-8")
	header.Set("Cache-Control", "no-cache")
	request.Response.WriteHeader(http.StatusOK)
	_, err = request.Response.Write(data)
	return err
}
